File: julia/src/benchmarks/Fannkuchredux.jl
using StaticArrays

mutable struct Fannkuchredux <: AbstractBenchmark
    n::Int64
    result::UInt32

    function Fannkuchredux()
        n = Helper.config_i64("Fannkuchredux", "n")
        new(n, UInt32(0))
    end
end

name(b::Fannkuchredux)::String = "Fannkuchredux"

function run(b::Fannkuchredux, iteration_id::Int64)
    n = Int32(b.n)

    perm1 = MVector{16, Int32}(undef)
    perm = MVector{16, Int32}(undef)
    count = MVector{16, Int32}(undef)

    i = Int32(0)
    while i < n
        perm1[i+1] = i  
        perm[i+1] = 0
        count[i+1] = 0
        i += 1
    end

    max_flips = Int32(0)
    perm_count = Int32(0)
    checksum = Int32(0)
    r = n

    while true

        while r > 1
            count[r] = r  
            r -= 1
        end

        i = Int32(0)
        while i < n
            perm[i+1] = perm1[i+1]
            i += 1
        end

        flips_count = Int32(0)
        k = perm[1]  

        while k != 0
            k2 = (k + 1) >> 1  

            i_local = Int32(0)
            while i_local < k2
                j = k - i_local

                temp = perm[i_local+1]
                perm[i_local+1] = perm[j+1]
                perm[j+1] = temp
                i_local += 1
            end

            flips_count += 1
            k = perm[1]  
        end

        if flips_count > max_flips
            max_flips = flips_count
        end

        if (perm_count & 1) == 0
            checksum += flips_count
        else
            checksum -= flips_count
        end

        while true

            if r == n

                result_int64 = (Int64(checksum) * 100 + Int64(max_flips)) & 0xffffffff
                b.result += UInt32(result_int64)
                return
            end

            perm0 = perm1[1]  

            i = Int32(0)
            while i < r
                perm1[i+1] = perm1[i+2]  
                i += 1
            end

            perm1[r+1] = perm0  

            count[r+1] -= 1  
            if count[r+1] > 0
                break
            end
            r += 1
        end

        perm_count += 1
    end
end

checksum(b::Fannkuchredux)::UInt32 = b.result

File: kotlin/src/main/kotlin/benchmarks/Fannkuchredux.kt
package benchmarks

import Benchmark

class Fannkuchredux : Benchmark() {
    private var n: Long = 0
    private var resultVal: UInt = 0u

    init {
        n = configVal("n")
    }

    private data class Result(
        val checksum: Int,
        val maxFlipsCount: Int,
    )

    private fun fannkuchredux(n: Int): Result {
        val perm1 = IntArray(32) { it }
        val perm = IntArray(32)
        val count = IntArray(32)
        var maxFlipsCount = 0
        var permCount = 0
        var checksum = 0
        var r = n

        while (true) {
            while (r > 1) {
                count[r - 1] = r
                r -= 1
            }

            System.arraycopy(perm1, 0, perm, 0, n)

            var flipsCount = 0
            var k = perm[0]

            while (k != 0) {
                val k2 = (k + 1) shr 1
                for (i in 0 until k2) {
                    val j = k - i
                    val temp = perm[i]
                    perm[i] = perm[j]
                    perm[j] = temp
                }
                flipsCount += 1
                k = perm[0]
            }

            if (flipsCount > maxFlipsCount) {
                maxFlipsCount = flipsCount
            }

            checksum += if (permCount % 2 == 0) flipsCount else -flipsCount

            while (true) {
                if (r == n) {
                    return Result(checksum, maxFlipsCount)
                }

                val perm0 = perm1[0]
                for (i in 0 until r) {
                    val j = i + 1
                    val temp = perm1[i]
                    perm1[i] = perm1[j]
                    perm1[j] = temp
                }

                perm1[r] = perm0
                count[r] -= 1
                val cntr = count[r]
                if (cntr > 0) break
                r += 1
            }
            permCount += 1
        }
    }

    override fun run(iterationId: Int) {
        val (a, b) = fannkuchredux(n.toInt())
        resultVal += (a * 100 + b).toUInt()
    }

    override fun checksum(): UInt = resultVal

    override fun name(): String = "CLBG::Fannkuchredux"
}
